### PR TITLE
SDS-135: Parse a GDPR Retention Policy

### PR DESCRIPTION
**Pull request by papamaemae**
## Description of change

To make the SDS API GDPR-compliant, we want to support defining a retention policy - a delta of time after which a file should be marked for deletion. This PR implements a retention period parser, that calculates the expiry date for a file based on a start day and a provided retention policy.

The retention policy is a string, and must be formatted with a suffix of `y`, `m` or `d` to represent years, months or days (e.g. `10y` or `56d`).

Specific named errors are raised if retention period is the value `DO NOT DELETE` (valid for files that should never be marked for deletion) or the value UNKNOWN (we want to avoid this, files should always be marked with a retention policy or `DO NOT DELETE`).

The logic accounts for leap years and negative values.

Unit tested for the following cases:
Valid inputs: '10y', '6m', '30d'
Edge cases: '0d', '999y'
Invalid formats: '10x', 'abc', '10', '-10y'
Special cases: 'DO NOT DELETE', 'UNKNOWN'

Postman/Pytest end-to-end tests have not been updated as the parser is currently not consumed anywhere in the app and does not affect any of our e2e journeys.

## Link to Jira Ticket

- [SDS-135](https://dsdmoj.atlassian.net/browse/SDS-{ticketNumber})

## Screenshots or test evidence if applicable

Unit tests all pass.

Screenshot 2025-09-01 at 15 57 12

### CODE REVIEW

**Review by alanmaydwell — CHANGES_REQUESTED** [review]
An interesting project.  Good to see this can be done in a concise way with clear explanation in docstring. 

I've left a few comments but the only one that I think needs a change is the Pipfile one.

The other comments don't matter so much, especially as this is an initial version but there are plans to later expand to cover named durations like `seven years`.

File: tests/utils/test_retention_policy_parser.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
+import unittest
```
[the remark this answers is not in the document]
Comment (papamaemae) [reply tests/utils/test_retention_policy_parser.py:1]: I actually meant to refactor this to the pytest framework for consistency against the rest of our tests. I got used to Python's built in unittest framework when starting to learn Python so my brain apparently defaulted to using that when writing these tests, but pytest is more modern and flexible. Have now refactored :)
